// @sap/approuter/lib/utils/app-cache-utils.js
'use strict';

const APPS_CACHE_SESSION_PREFIX = '{appsCacheSession}';
const APPS_CACHE_LOCK_PREFIX = '{appsCacheLock}';
const MAX_CACHE_SIZE = 200 * 1024; // 200KB
const { v4: uuidv4 } = require('uuid');

module.exports.setAppsCache = async function(cacheKey, cacheValue, req){
  if (!isAppsCacheEnabled(req)) {return;}
  const tracer = req.loggingContext && req.loggingContext.getTracer(__filename);
  const cacheValueSize = cacheValue && JSON.stringify(cacheValue).length;
  if (cacheValueSize > MAX_CACHE_SIZE) {
    tracer && tracer.error(`Maximum cache value size of ${MAX_CACHE_SIZE} exceeded for cache key ${cacheKey}`);
    return;
  }
  const appCacheSession = {
    externalSessionId: cacheKey,
    externalSessionExpiration: getSessionExpiration(req),
    applications: cacheValue
  };
  tracer && tracer.debug(`Storing apps cache for cache key : ${cacheKey}`);
  await req.app.get('externalSessionStore').updateExternalSession(appCacheSession, APPS_CACHE_SESSION_PREFIX);
};

module.exports.getAppsCache = async function(cacheKey, req){
  if (!isAppsCacheEnabled(req)) {return;}
  const tracer = req.loggingContext && req.loggingContext.getTracer(__filename);
  const appCacheSession = await req.app.get('externalSessionStore').getExternalSession(cacheKey,null, APPS_CACHE_SESSION_PREFIX);
  if (appCacheSession && appCacheSession.applications){
    tracer && tracer.debug(`Retrieved apps cache for cache key : ${cacheKey}`);
    return appCacheSession.applications;
  }
  tracer && tracer.debug(`Apps cache not found for cache key : ${cacheKey}`);
  return null;
};

module.exports.deleteAppsCache = async function(cacheKey, req){
  if (!isAppsCacheEnabled(req)) {return;}
  const tracer = req.loggingContext && req.loggingContext.getTracer(__filename);
  tracer && tracer.debug(`Deleting apps cache for cache key : ${cacheKey}`);
  await req.app.get('externalSessionStore').destroyExternalSession(cacheKey, APPS_CACHE_SESSION_PREFIX);
};

function getSessionExpiration(req){
  const routerConfig = req.app.get('mainRouterConfig');
  const sessionLength = routerConfig.appsCacheExpiration;
  const tracer = req.loggingContext && req.loggingContext.getTracer(__filename);
  const appSessionExpiration = (Math.floor(Date.now() / 1000)) + sessionLength;
  tracer && tracer.debug(`Apps cache session expiration: ${appSessionExpiration}`);
  return appSessionExpiration;
}

function isAppsCacheEnabled(req){
  const routerConfig = req.app.get('mainRouterConfig');
  return routerConfig && routerConfig.appsCacheExpiration > 0;
}

module.exports.acquireLock = async function(lockKey, lockTTL, req) {
  const tracer = req.loggingContext && req.loggingContext.getTracer(__filename);
  tracer && tracer.debug(`Requesting acquireLock for key: ${lockKey} with TTL: ${lockTTL}ms`);

  const store = req.app.get('externalSessionStore');
  const token = uuidv4();

  // Check if lock exists
  const existingLock = await store.getExternalSession(lockKey, null, APPS_CACHE_LOCK_PREFIX);
  if (existingLock) {
    return { acquired: false };
  }

  // Create lock session with expiration timestamp
  const lockSession = {
    externalSessionId: lockKey,
    externalSessionExpiration: Math.floor(Date.now() / 1000) + (lockTTL / 1000),
    token: token
  };

  // Determine if we got the lock
  const acquired = await store.updateExternalSession(lockSession, APPS_CACHE_LOCK_PREFIX, true);

  return { acquired, token: acquired ? token : null };
};

module.exports.renewLock = async function(lockKey, token, lockTTL, req) {
  const tracer = req.loggingContext && req.loggingContext.getTracer(__filename);
  tracer && tracer.debug(`Requesting renewLock for key: ${lockKey} with token: ${token} and TTL: ${lockTTL}ms`);

  const store = req.app.get('externalSessionStore');

  // Check if we own the lock
  const currentLock = await store.getExternalSession(lockKey, null, APPS_CACHE_LOCK_PREFIX);
  if (!currentLock || currentLock.token !== token) {
    return false;
  }

  // Update lock with new expiration
  const lockSession = {
    externalSessionId: lockKey,
    externalSessionExpiration: Math.floor(Date.now() / 1000) + (lockTTL / 1000),
    token: token
  };
  await store.updateExternalSession(lockSession, APPS_CACHE_LOCK_PREFIX);

  return true;
};

module.exports.releaseLock = async function(lockKey, token, req) {
  const tracer = req.loggingContext && req.loggingContext.getTracer(__filename);
  tracer && tracer.debug(`Requesting releaseLock for key: ${lockKey} with token: ${token}`);

  const store = req.app.get('externalSessionStore');

  // Check if we own the lock
  const currentLock = await store.getExternalSession(lockKey, null, APPS_CACHE_LOCK_PREFIX);
  if (currentLock && currentLock.token === token) {
    await store.destroyExternalSession(lockKey, APPS_CACHE_LOCK_PREFIX);
  }
};
